protocols/mdns/src/instance.rs

```diff
@@ -0,0 +1,228 @@
+    }
+
+    fn inject_mdns_packet(&mut self, packet: MdnsPacket, params: &impl PollParameters) {
+        log::trace!("received packet on iface {} {:?}", self.addr, packet);
+        match packet {
+            MdnsPacket::Query(query) => {
+                self.reset_timer();
+                log::trace!("sending response on iface {}", self.addr);
+                for packet in build_query_response(
+                    query.query_id(),
+                    *params.local_peer_id(),
+                    params.listened_addresses(),
+                    self.ttl,
+                ) {
+                    self.send_buffer.push_back(packet);
+                }
+            }
+            MdnsPacket::Response(response) => {
+                // We replace the IP address with the address we observe the
+                // remote as and the address they listen on.
+                let obs_ip = Protocol::from(response.remote_addr().ip());
+                let obs_port = Protocol::Udp(response.remote_addr().port());
+                let observed: Multiaddr = iter::once(obs_ip).chain(iter::once(obs_port)).collect();
+
+                for peer in response.discovered_peers() {
+                    if peer.id() == params.local_peer_id() {
+                        continue;
+                    }
+
+                    let new_expiration = Instant::now() + peer.ttl();
+
+                    let mut addrs: Vec<Multiaddr> = Vec::new();
+                    for addr in peer.addresses() {
+                        if let Some(new_addr) = address_translation(addr, &observed) {
+                            addrs.push(new_addr.clone())
+                        }
+                        addrs.push(addr.clone())
+                    }
+
+                    for addr in addrs {
+                        self.discovered
+                            .push_back((*peer.id(), addr, new_expiration));
+                    }
+                }
+            }
+            MdnsPacket::ServiceDiscovery(disc) => {
+                let resp = build_service_discovery_response(disc.query_id(), self.ttl);
+                self.send_buffer.push_back(resp);
+            }
+        }
+    }
+
+    pub fn poll(
+        &mut self,
+        cx: &mut Context,
+        params: &impl PollParameters,
+    ) -> Option<(PeerId, Multiaddr, Instant)> {
+        // Poll receive socket.
+        while self.recv_socket.poll_readable(cx).is_ready() {
+            match self
+                .recv_socket
+                .recv_from(&mut self.recv_buffer)
+                .now_or_never()
+            {
+                Some(Ok((len, from))) => {
+                    if let Some(packet) = MdnsPacket::new_from_bytes(&self.recv_buffer[..len], from)
+                    {
+                        self.inject_mdns_packet(packet, params);
+                    }
+                }
+                Some(Err(err)) => log::error!("Failed reading datagram: {}", err),
+                None => unreachable!(),
```

Comment:
So this is impossible because we call is_ready() before?

Reply:
exactly. although can revert as I was just trying to rule out some code paths during debugging